lnwire/accept_channel.go

```diff
@@ -150,15 +171,82 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
 		return err
 	}
 
-	// Check for the optional upfront shutdown script field. If it is not there,
-	// silence the EOF error.
-	err = ReadElement(r, &a.UpfrontShutdownScript)
-	if err != nil && err != io.EOF {
+	// For backwards compatibility, the optional extra data blob for
+	// AcceptChannel must contain an entry for the upfront shutdown script.
+	// We'll read it out and attempt to parse it.
+	var tlvRecords ExtraOpaqueData
+	if err := ReadElements(r, &tlvRecords); err != nil {
 		return err
 	}
+
+	a.UpfrontShutdownScript, a.ExtraData, err = parseShutdownScript(
```

Comment:
i think it makes sense to use `ExtraOpaqueData` deserialization for those cases where we don't have any defined TLV fields, but in cases where we have known fields I had imagined using a custom `tlv.Stream` that would  container a pointer to the fields that need to be [de]serialized. In the long run, i don't think it will be scalable to have a `parseXXX` field for everything we want to end up extracting

Reply:
What about something like:

```
// Assumes all records being parsed start at the beginning of the TLV stream and follow one another.
// If this isn't the case, we'd need the custom `tlv.Stream` that contains pointers to each record.
func (e *ExtraOpaqueData) ParseKnownRecords(producers ...tlv.RecordProducer) error {
	for _, p := range producers {
		r := p.Record()
		records, err := e.ExtractRecords(r)
		if err != nil {
			return err
		}
		rawRecordVal, ok := records[r.Type()]
		if !ok || rawRecordVal != nil {
			return fmt.Errorf("record with type %v not parsed",
				r.Type())
		}

		// Remove the bytes corresponding to the record we just parsed
		// from our TLV stream.
		parsedRecord := p.Record()
		valSize := parsedRecord.Size()
		lenSize := tlv.VarIntSize(valSize)
		typeSize := tlv.VarIntSize(uint64(parsedRecord.Type()))
		*e = []byte(*e)[typeSize+lenSize+valSize:]
	}

	return nil
}
```

In the case of `lnwire.OpenChannel`, it'd be used as:

```
	var tlvRecords ExtraOpaqueData
	if err := ReadElements(r, &tlvRecords); err != nil {
		return err
	}
	err = tlvRecords.ParseKnownRecords(&o.UpfrontShutdownScript)
	if err != nil {
		return err
	}
	o.ExtraData = tlvRecords
```

Reply:
I agree that this approach won't be scalable in the generic case, but the shutdown script is really a _special case_, since it must always be there. 

I think we should defer making this more generic until we have actual new fields we want to add, we'll know more about the requirements then.

Reply:
Agreed, we can move ahead with the bulk of this PR given once we are properly writing/parsing all the data, we can make any future in-memory TLV representation changes at a later point.